File: registry-compiler/src/main/java/com/github/fengdai/registry/compiler/RegisterAnnotation.java
package com.github.fengdai.registry.compiler;

import com.google.auto.common.MoreTypes;
import com.google.common.collect.ImmutableSet;
import com.sun.tools.javac.code.Attribute;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.TypeElement;

import static com.google.auto.common.AnnotationMirrors.getAnnotationValue;

final class RegisterAnnotation {
  final AnnotationMirror annotation;

  RegisterAnnotation(AnnotationMirror annotation) {
    this.annotation = annotation;
  }

  ImmutableSet<TypeElement> binders() {
    return ((Attribute.Array) getAnnotationValue(annotation, "binders")).getValue()
        .stream()
        .filter(attribute -> attribute instanceof Attribute.Class)
        .map(it -> MoreTypes.asTypeElement(((Attribute.Class) it).getValue()))
        .collect(ImmutableSet.toImmutableSet());
  }

  ImmutableSet<TypeElement> binderViewHolders() {
    return ((Attribute.Array) getAnnotationValue(annotation, "binderViewHolders")).getValue()
        .stream()
        .filter(attribute -> attribute instanceof Attribute.Class)
        .map(it -> MoreTypes.asTypeElement(((Attribute.Class) it).getValue()))
        .collect(ImmutableSet.toImmutableSet());
  }

  ImmutableSet<Integer> staticContentLayoutsValues() {
    return ((Attribute.Array) getAnnotationValue(annotation, "staticContentLayouts")).getValue()
        .stream()
        .map(it -> (Integer) ((Attribute.Constant) it).value)
        .collect(ImmutableSet.toImmutableSet());
  }
}

File: registry-processor/src/main/java/com/github/fengdai/registry/processor/RegisterAnnotation.kt
package com.github.fengdai.registry.processor

import com.google.auto.common.AnnotationMirrors.getAnnotationValue
import com.google.auto.common.MoreTypes
import com.sun.tools.javac.code.Attribute
import javax.lang.model.element.AnnotationMirror
import javax.lang.model.element.TypeElement
import javax.lang.model.type.TypeMirror

data class RegisterAnnotation(val annotation: AnnotationMirror) {
  val binders: Set<TypeElement> =
    (getAnnotationValue(annotation, "binders") as Attribute.Array).value
        .asSequence()
        .filterIsInstance<Attribute.Class>()
        .map { it.value.asTypeElement() }
        .toSet()

  val binderViewHolders: Set<TypeElement> =
    (getAnnotationValue(annotation, "binderViewHolders") as Attribute.Array).value
        .filterIsInstance<Attribute.Class>()
        .map { it.value.asTypeElement() }
        .toSet()

  val staticContentLayoutsValues: Set<Int> =
    (getAnnotationValue(annotation, "staticContentLayouts") as Attribute.Array).value
        .map { (it as Attribute.Constant).value as Int }
        .toSet()
}

private fun TypeMirror.asTypeElement() = MoreTypes.asTypeElement(this)
